src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -142,7 +141,16 @@ private void LowerDecisionDag(ImmutableArray<BoundDecisionDagNode> sortedNodes)
                         break;
                 }
 
-                // Code for the when clause goes in a separate code section for the switch section.
+                // Note that a when-clause can contain an assignment to a
+                // pattern variable declared in a different when-clause, so we only apply the optimization
+                // of using user-declared pattern variables as pattern-matching automaton temps
+                // when there are no nontrivial when-clauses at all.
+                if (!sortedNodes.OfType<BoundWhenDecisionDagNode>().Where(w => w.WhenExpression != null && w.WhenExpression.ConstantValue != ConstantValue.True).Any())
```

Comment:
I think we could do slightly better here by tracking inside the when clauses to see if a variable is modified or calls something that could modify that variable (local function, etc). This would be rare, but when clauses are probably not so rare.